pkg/pool-hooks/test/foundry/StableSurgeHookTest.sol

```diff
@@ -0,0 +1,336 @@
+
+        authorizer.grantRole(vault.getActionId(StableSurgeHook.setSurgeThresholdPercentage.selector), address(this));
+
+        vm.expectEmit();
+        emit StableSurgeHook.ThresholdSurgePercentageChanged(pool, newSurgeThresholdPercentage);
+        stableSurgeHook.setSurgeThresholdPercentage(pool, newSurgeThresholdPercentage);
+
+        assertEq(
+            stableSurgeHook.getSurgeThresholdPercentage(pool),
+            newSurgeThresholdPercentage,
+            "Surge threshold percentage should be newSurgeThresholdPercentage"
+        );
+    }
+
+    function testChangeSurgeThresholdPercentageRevertIfValueIsGreaterThanOne() public {
+        uint256 newSurgeThresholdPercentage = 1.1e18;
+
+        PoolRoleAccounts memory poolRoleAccounts = PoolRoleAccounts({
+            pauseManager: address(this),
+            swapFeeManager: address(this),
+            poolCreator: address(this)
+        });
+
+        vm.mockCall(
+            address(vault),
+            abi.encodeWithSelector(IVaultExplorer.getPoolRoleAccounts.selector, pool),
+            abi.encode(poolRoleAccounts)
+        );
+
+        vm.expectRevert(StableSurgeHook.InvalidSurgeThresholdPercentage.selector);
+        stableSurgeHook.setSurgeThresholdPercentage(pool, newSurgeThresholdPercentage);
+    }
+
+    function testChangeSurgeThresholdPercentageRevertIfSenderIsNotFeeManagerAndNotAuthorizer() public {
+        vm.expectRevert(IAuthentication.SenderNotAllowed.selector);
+
+        stableSurgeHook.setSurgeThresholdPercentage(pool, 1e18);
+    }
+
+    function testGetSurgeFeePercentage() public view {
+        uint256 numTokens = 8;
+        uint256[] memory balances = new uint256[](numTokens);
+
+        uint256 amountGivenScaled18 = 1e18;
+        uint256 tokenIn = 0;
+        uint256 tokenOut = 7;
+
+        balances[0] = 1e18;
+        balances[1] = 1e18;
+        balances[2] = 1e18;
+        balances[3] = 10000e18;
+        balances[4] = 2e18;
+        balances[5] = 3e18;
+        balances[6] = 1e18;
+        balances[7] = 1e18;
+
+        PoolSwapParams memory params = PoolSwapParams({
+            kind: SwapKind.EXACT_IN,
+            indexIn: tokenIn,
+            indexOut: tokenOut,
+            amountGivenScaled18: amountGivenScaled18,
+            balancesScaled18: balances,
+            router: address(0),
+            userData: bytes("")
+        });
+        uint256 surgeThresholdPercentage = DEFAULT_SURGE_THRESHOLD_PERCENTAGE;
+        uint256 staticFeePercentage = 1e16;
+
+        uint256 surgeFeePercentage = stableSurgeHook.getSurgeFeePercentage(
+            params,
+            surgeThresholdPercentage,
+            staticFeePercentage
+        );
+
+        uint256[] memory newBalances = new uint256[](numTokens);
+        for (uint256 i = 0; i < numTokens; ++i) {
+            newBalances[i] = balances[i];
+        }
+        newBalances[tokenIn] += amountGivenScaled18;
+        newBalances[tokenOut] -= amountGivenScaled18;
```

Comment:
This calculation is very rough since a stable pool uses a curve, and the input amount does not have a 1:1 ratio with the output amount. I don’t see any other way, except for a highly suboptimal approach where we replace these calculations with actual output amount calculations using `calculateInvariant` and `calculateBalances`. However, in terms of gas costs, this would be equivalent to performing a full swap calculation, which seems impractical. WDYT?

@jubeira @EndymionJkb

Reply:
I think the only correct option is to call `pool.onSwap` inside the hook. This will increase gas usage, but maybe not significantly. In other words, I would rely here on whether the current rough calculations break the logic of the Surgery Hook or not.

Reply:
I get the feeling that we'd need to analyze this a bit further.

For a 2-pool token, the median is the average.
If you consider that the change is going to be 1:1, the median doesn't change after the operation. E.g. you start with 100, 200 tokens, add X. Original average is 150, new average is `[(100 + x) + (200 - x)] / 2 = 150`.

If you apply the trading curve that shouldn't be the case in the extremes, where you have to add a lot of tokens A to get just a few tokens B.

Then, that median is used to compute the imbalance.
I'm not saying that the current approach is wrong, but sounds like the effect will be quite different from the previous approach.

Reply:
@EndymionJkb did the calculation in excel and we came to the conclusion that median works better (if we talk about the case of more than 2 tokens)